pkg/deploy/predeploy.go

```diff
@@ -0,0 +1,268 @@
+}
+
+func (d *deployer) deployPreDeploy(ctx context.Context, rpServicePrincipalID string) error {
+	deploymentName := "rp-production-predeploy"
+
+	_, err := d.deployments.Get(ctx, d.config.ResourceGroupName, deploymentName)
+	if err == nil || !isDeploymentNotFoundError(err) {
+		return err
+	}
+
+	b, err := Asset(generator.FileRPProductionPredeploy)
+	if err != nil {
+		return err
+	}
+
+	var template map[string]interface{}
+	err = json.Unmarshal(b, &template)
+	if err != nil {
+		return err
+	}
+
+	parameters := d.getParameters(template["parameters"].(map[string]interface{}))
+	parameters.Parameters["rpServicePrincipalId"] = &arm.ParametersParameter{
+		Value: rpServicePrincipalID,
+	}
+
+	d.log.Infof("predeploying to %s", d.config.ResourceGroupName)
+	return d.deployments.CreateOrUpdateAndWait(ctx, d.config.ResourceGroupName, deploymentName, mgmtresources.Deployment{
+		Properties: &mgmtresources.DeploymentProperties{
+			Template:   template,
+			Mode:       mgmtresources.Incremental,
+			Parameters: parameters.Parameters,
+		},
+	})
+}
+
+func (d *deployer) configureServiceKV(ctx context.Context) error {
+	serviceKeyVaultURI := "https://" + d.config.Configuration.KeyvaultPrefix + "-svc.vault.azure.net/"
+
+	err := d.ensureEncryptionSecret(ctx, serviceKeyVaultURI)
+	if err != nil {
+		return err
+	}
+
+	err = d.ensureMonitoringCertificates(ctx, serviceKeyVaultURI)
+	if err != nil {
+		return err
+	}
+
+	return d.ensureServiceCertificates(ctx, serviceKeyVaultURI)
+}
+
+func (d *deployer) ensureEncryptionSecret(ctx context.Context, serviceKeyVaultURI string) error {
+	secrets, err := d.keyvault.GetSecrets(ctx, serviceKeyVaultURI, nil)
+	if err != nil {
+		return err
+	}
+
+	for _, secret := range secrets {
+		if filepath.Base(*secret.ID) == env.EncryptionSecretName {
+			return nil
+		}
+	}
+
+	key := make([]byte, 32)
+	_, err = rand.Read(key)
+	if err != nil {
+		return err
+	}
+
+	_, err = d.keyvault.SetSecret(ctx, serviceKeyVaultURI, env.EncryptionSecretName, keyvault.SecretSetParameters{
+		Value: to.StringPtr(string(key)),
+	})
+	return err
+}
+
+func (d *deployer) ensureMonitoringCertificates(ctx context.Context, serviceKeyVaultURI string) error {
+	for _, certificateName := range []string{
+		env.ClusterLoggingSecretName,
+		env.RPLoggingSecretName,
+		env.RPMonitoringSecretName,
+	} {
+		bundle, err := d.keyvault.GetSecret(ctx, d.config.Configuration.GlobalMonitoringKeyVaultURI, certificateName, "")
+		if err != nil {
+			return err
+		}
+
+		_, err = d.keyvault.ImportCertificate(ctx, serviceKeyVaultURI, certificateName, keyvault.CertificateImportParameters{
+			Base64EncodedCertificate: bundle.Value,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *deployer) ensureServiceCertificates(ctx context.Context, serviceKeyVaultURI string) error {
+	_, err := d.keyvault.SetCertificateIssuer(ctx, serviceKeyVaultURI, "OneCert", keyvault.CertificateIssuerSetParameters{
```

Comment:
Will this work during Pre-Deploy? I understood that this is only works after keyvault is being enable to work with OneCert?

Reply:
see other comment about OneCert cert generation works